File: cli/internal/crypto/aes.go
// Package crypto provides AES-256-GCM encryption and PBKDF2 key derivation
// for the MarkPush protocol.
//
// Ciphertext format: nonce (12 bytes) || ciphertext || GCM auth tag (16 bytes),
// then base64-standard-encoded. Both CLI and iOS must use this same format.
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
	"io"

	"golang.org/x/crypto/pbkdf2"
)

const (
	// KeySize is the required key length in bytes (256 bits).
	KeySize = 32
	// PBKDF2Iterations is the number of PBKDF2 iterations for key derivation.
	PBKDF2Iterations = 100_000
)

// Encrypt encrypts plaintext with AES-256-GCM.
// Key must be exactly 32 bytes. Returns base64-encoded ciphertext.
func Encrypt(key, plaintext []byte) (string, error) {
	if len(key) != KeySize {
		return "", fmt.Errorf("encrypt: key must be %d bytes, got %d", KeySize, len(key))
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("encrypt: create cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("encrypt: create GCM: %w", err)
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", fmt.Errorf("encrypt: generate nonce: %w", err)
	}

	// Seal appends ciphertext+tag to nonce, producing: nonce || ciphertext || tag
	ciphertext := gcm.Seal(nonce, nonce, plaintext, nil)
	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

// Decrypt decrypts a base64-encoded AES-256-GCM ciphertext.
// Key must be exactly 32 bytes. The encoded string must contain
// the nonce prepended to the ciphertext.
func Decrypt(key []byte, encoded string) ([]byte, error) {
	if len(key) != KeySize {
		return nil, fmt.Errorf("decrypt: key must be %d bytes, got %d", KeySize, len(key))
	}

	data, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return nil, fmt.Errorf("decrypt: base64 decode: %w", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("decrypt: create cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("decrypt: create GCM: %w", err)
	}

	nonceSize := gcm.NonceSize()
	if len(data) < nonceSize {
		return nil, errors.New("decrypt: ciphertext too short")
	}

	nonce, ciphertext := data[:nonceSize], data[nonceSize:]
	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("decrypt: authentication failed: %w", err)
	}

	return plaintext, nil
}

// DeriveKey derives a 32-byte encryption key from a pairing secret and salt
// using PBKDF2 with SHA-256.
func DeriveKey(secret string, salt []byte) ([]byte, error) {
	if secret == "" {
		return nil, errors.New("derive key: secret must not be empty")
	}
	if len(salt) == 0 {
		return nil, errors.New("derive key: salt must not be empty")
	}

	key := pbkdf2.Key([]byte(secret), salt, PBKDF2Iterations, KeySize, sha256.New)
	return key, nil
}

File: mcp/src/crypto/aes.ts
/**
 * AES-256-GCM encryption/decryption using Node.js crypto module.
 *
 * Ciphertext format: nonce (12 bytes) || ciphertext || GCM auth tag (16 bytes),
 * then base64-standard-encoded. Must match the Go CLI and Swift iOS implementations.
 */

import { createCipheriv, createDecipheriv, randomBytes, pbkdf2Sync } from "node:crypto";

const KEY_SIZE = 32;
const NONCE_SIZE = 12;
const TAG_SIZE = 16;
const PBKDF2_ITERATIONS = 100_000;

/** Encrypt plaintext with AES-256-GCM. Returns base64-encoded ciphertext. */
export function encrypt(key: Buffer, plaintext: Buffer): string {
  if (key.length !== KEY_SIZE) {
    throw new Error(`Key must be ${KEY_SIZE} bytes, got ${key.length}`);
  }

  const nonce = randomBytes(NONCE_SIZE);
  const cipher = createCipheriv("aes-256-gcm", key, nonce);

  const encrypted = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();

  // Format: nonce || ciphertext || tag
  const combined = Buffer.concat([nonce, encrypted, tag]);
  return combined.toString("base64");
}

/** Decrypt base64-encoded AES-256-GCM ciphertext. */
export function decrypt(key: Buffer, encoded: string): Buffer {
  if (key.length !== KEY_SIZE) {
    throw new Error(`Key must be ${KEY_SIZE} bytes, got ${key.length}`);
  }

  const data = Buffer.from(encoded, "base64");
  if (data.length < NONCE_SIZE + TAG_SIZE) {
    throw new Error("Ciphertext too short");
  }

  const nonce = data.subarray(0, NONCE_SIZE);
  const ciphertext = data.subarray(NONCE_SIZE, data.length - TAG_SIZE);
  const tag = data.subarray(data.length - TAG_SIZE);

  const decipher = createDecipheriv("aes-256-gcm", key, nonce);
  decipher.setAuthTag(tag);

  return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
}

/**
 * Derive a 32-byte key from a pairing secret using PBKDF2-SHA256.
 * Parameters match CLI and iOS: 100,000 iterations, 32-byte output.
 */
export function deriveKey(secret: string, salt: string): Buffer {
  if (!secret) throw new Error("Secret must not be empty");
  if (!salt) throw new Error("Salt must not be empty");

  return pbkdf2Sync(secret, salt, PBKDF2_ITERATIONS, KEY_SIZE, "sha256");
}
